[Lens] Add standard deviation function

flash1293 opened the pull request:
Fixes https://github.com/elastic/kibana/issues/10533

This PR adds the standard deviation function to Lens.

## AggConfig changes

To the existing `std_deviation` agg config type which returns the upper and lower bound, add an argument "showBounds" which can be set to false to return a regular single metric and return the regular standard deviation.

## Lens changes

Expose in quick function list and formula:
<img width="354" alt="Screenshot 2022-07-13 at 18 29 38" src="https://user-images.githubusercontent.com/1508364/178790044-bf8b8352-083f-464e-bd2c-f7d32c00855b.png">
<img width="385" alt="Screenshot 2022-07-13 at 18 30 07" src="https://user-images.githubusercontent.com/1508364/178790048-06564e8f-ce8f-4484-b77d-c6a492d6998c.png">
<img width="1013" alt="Screenshot 2022-07-13 at 19 00 58" src="https://user-images.githubusercontent.com/1508364/178790051-3ae40702-3026-4d15-be7b-940c95ec0bda.png">
<img width="948" alt="Screenshot 2022-07-13 at 19 07 36" src="https://user-images.githubusercontent.com/1508364/178791027-4e3cc274-8164-4b5b-9df9-c85bbb365e76.png">

It's mentioned in the issue but I didn't add sum of squares and variance as individual quick function entries - these seem too rare to put into this highly visible place and they can also be achieved using standard deviation and formula. I put them into the documentation as examples

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

ppisljar reviewed (approved):
app services code LGTM

KOTungseth commented:
The changes in the screenshots look great! I have some minor copy changes for the following:

<img width="1013" alt="178790051-3ae40702-3026-4d15-be7b-940c95ec0bda" src="https://user-images.githubusercontent.com/38990715/179047993-beff8bd6-9d70-4faf-a984-a6e3cc132896.png">

- We should include the definition of standard deviation
- Instead of using `Example` three times in a row, let's place all the examples under an `Example` heading. 

I propose the following:
**NOTE: I don't love `squares of price`, but I'm not sure what else to use here.**

standard_deviation (field: string)

Returns the amount of variation or dispersion of the field. The function works only for number fields. 

Examples:

To get the standard deviation of price, use `standard_deviation(price)`.

To get the variance of prices for orders from the UK, use `square(standard_deviation(price, kql='location: UK'))`.

To get the sum of the squares of price, use `square(standard_deviation(price)) * count()`.

stratoula commented:
@flash1293 this works great but I found this
<img width="1470" alt="image" src="https://user-images.githubusercontent.com/17003240/179455034-24b7647c-7066-4386-ad41-c89618b79755.png">

It happens for partitions with top values, also with vertical bars with top values on the horizontal axis (with date histograms it works fine)
It works for agg-based visualizations so I guess that has to do with the aggId or something like that. I remember I had encountered the same problem when I was adding the percentile ranks agg.

flash1293 commented:
@stratoula Good catch, I forgot to define the `getValueBucketPath` method on the agg config which is used by the terms agg config to set the `order` property. Should work fine now.

flash1293 commented:
Adjusted documentation:
<img width="647" alt="Screenshot 2022-07-18 at 10 24 19" src="https://user-images.githubusercontent.com/1508364/179472320-ae742b95-7919-42de-bcdf-1b49d947fe3e.png">

I removed the sum of squares examples as it's an exotic edge case and can quickly get confusing in this context.

stratoula reviewed (approved):
Yes Joe, this works great now! LGTM!
